builder/common_builder/common_builder.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2020 Sorint.lab S.p.A.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package common
+
+import (
+	"runtime"
+	"strings"
+
+	"github.com/ercole-io/ercole-agent/config"
+	"github.com/ercole-io/ercole-agent/fetcher"
+	"github.com/ercole-io/ercole-agent/logger"
+	"github.com/ercole-io/ercole-agent/model"
+	"github.com/ercole-io/ercole-agent/utils"
+)
+
+// CommonBuilder for Linux and Windows hosts
+type CommonBuilder struct {
+	fetcher       fetcher.Fetcher
+	configuration config.Configuration
+	log           logger.Logger
+}
+
+// NewCommonBuilder initialize an appropriate builder for Linux or Windows
+func NewCommonBuilder(configuration config.Configuration, log logger.Logger) CommonBuilder {
+	var specializedFetcher fetcher.SpecializedFetcher
+
+	log.Debugf("runtime.GOOS: [%v]", runtime.GOOS)
+
+	if runtime.GOOS == "windows" {
+		wf := fetcher.NewWindowsFetcherImpl(configuration, log)
+		specializedFetcher = &wf
+
+	} else {
+		if runtime.GOOS != "linux" {
+			log.Errorf("Unknow runtime.GOOS: [%v], I'll try with linux\n", runtime.GOOS)
+		}
+
+		lf := fetcher.NewLinuxFetcherImpl(configuration, log)
+		specializedFetcher = &lf
+	}
+
+	fetcherImpl := &fetcher.CommonFetcherImpl{
+		SpecializedFetcher: specializedFetcher,
+		Configuration:      configuration,
+		Log:                log,
+	}
+
+	builder := CommonBuilder{
+		fetcher:       fetcherImpl,
+		configuration: configuration,
+		log:           log,
+	}
+
+	return builder
+}
+
+// Run fill hostData
+func (b *CommonBuilder) Run(hostData *model.HostData) {
+	if b.configuration.Features.Exadata.Enabled {
+		b.log.Debug("Exadata mode enabled")
+
+		b.checksToRunExadata()
+
+		if err := b.fetcher.SetUser(b.configuration.Features.Exadata.NotRootUser); err != nil {
+			b.log.Panicf("Can't set user [%s] for fetcher, err: [%v]", b.configuration.Features.Exadata.NotRootUser, err)
+		}
+	}
+
+	hostData.Info = *b.getHost()
+
+	hostData.Hostname = hostData.Info.Hostname
+	if b.configuration.Hostname != "default" {
+		hostData.Hostname = b.configuration.Hostname
+	}
+
+	if b.configuration.Features.Databases.Enabled {
+		b.log.Debug("Databases mode enabled")
+		hostData.Extra.Filesystems = b.fetcher.GetFilesystems()
+
+		hostData.Extra.Databases = b.getOracleDBs(hostData.Info.Type)
+		hostData.Databases, hostData.Schemas = b.getDatabasesAndSchemaNames(hostData.Extra.Databases)
+	}
```

Comment:
Why oracle don't use a NotRootUser? Oracle database fetchers should be executed with oracle or ercole user

Reply:
NotRootUser is set inside the fetcher: once it's set at line 77, every request run by that fetcher will use NotRootUser to execute command.

Reply:
In fact, before run b.getExadataDevices() I reset fetcher user to current user via b.fetcher.SetUserAsCurrent() at line 105

Reply:
It is setted only for exadata. I want to run ercole-agent with root permission and use the oracle fetcher with ercole account!

Reply:
Done :+1:

Reply:
Really???

Reply:
Pardon!
Run ercole-agent with root permission and use the oracle fetcher with ercole account isn't in the scope of this issue.
I can open another issue for that, and open another PR after this one!

Reply:
Ok...